### PR TITLE
Fixes null ref in DatabaseErrorPageMiddleware.

### PR DESCRIPTION
**Pull request by anpete**
Fix aspnet/Home#2820 - UseDatabaseErrorPage throws NullReferenceException when ef exception occur in background task
Fix aspnet/EntityFrameworkCore#9599 - Insert entity in non existing table throws NullException

Added null checks to async local access in event callback.

Related to aspnet/Home#2825

### CODE REVIEW

**Review by ajcvickers — APPROVED** [review]

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs [hunk 1]
```diff
@@ -216,15 +217,17 @@ void IObserver<DiagnosticListener>.OnNext(DiagnosticListener diagnosticListener)
         {
             switch (keyValuePair.Value)
             {
+                // NB: _localDiagnostic.Value can be null when this middleware has been leaked.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: can be null when DB operations take place outside of a request thread.

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs [hunk 2]
```diff
@@ -78,6 +78,8 @@ public void Hold(Exception exception, Type contextType)
             _options = options.Value;
             _logger = loggerFactory.CreateLogger<DatabaseErrorPageMiddleware>();
 
+            // Note: this currently leaks if the server hosting this middleware is disposed.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: Want to try the IMiddleware fix here?

**Review by Tratcher — COMMENTED** [review]
You opted not to try simplifying the design instead?

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs (hunk 2 above)
Comment (anpete) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: Depends what it is :trollface:

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs (hunk 1 above)
Comment (anpete) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: Want me to add that?

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs (hunk 2 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: Change the class to implement IMiddleware and IDisposable. The catch is that the class has to be registered as a service. Diagnostics does not currently register any services.

**Comment by anpete** [comment]
Re the design simplification, it turns out that any random database exception can end up here, and we need a way to deduce that it came from EF. Would love to improve this if you have any ideas.

**Comment by Tratcher** [comment]
Don't they all get wrapped?

**Comment by anpete** [comment]
No.

File: src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs (hunk 2 above)
Comment (anpete) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: What does registration of the middleware look like from UseDatabaseErrorPage?

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore/DatabaseErrorPageMiddleware.cs]: It's still `app.UseMiddleware<DatabaseErrorPageMiddleware>`, but passing in options directly might not work.

**Pull request merged**